File: exercises/floor02/ex01_booleans.py
"""
FLOOR 2 · EXERCISE 2-1 — Booleans & comparisons
Lesson: 2.1

Games are made of decisions, and decisions are made of booleans.
Comparisons (== != < > <= >=) produce True/False; and/or/not combine them.

Run:  python3 dojo.py check 2-1
Save: python3 dojo.py done 2-1
"""

player_hp = 7
player_max_hp = 30
has_key = True
door_locked = True
potions = 0
depth = 3
torch_lit = False

# --- Part 1: state checks ------------------------------------------------------
# is_hurt        -> True when hp is below max
# is_critical    -> True when hp is at or below 25% of max (use <=)
# TODO:
is_hurt = ...
is_critical = ...

# --- Part 2: combining ---------------------------------------------------------
# can_open_door  -> True when the player has the key AND the door is locked
#                   (opening an unlocked door needs no key logic)
# should_retreat -> True when critical AND out of potions
# TODO:
can_open_door = ...
should_retreat = ...

# --- Part 3: `not`, and the dark ------------------------------------------------
# in_darkness    -> True when the torch is NOT lit and depth is greater than 1
# TODO:
in_darkness = ...

# --- Part 4: spot the trap -------------------------------------------------------
# A previous adventurer wrote:   fatal = player_hp == 0 or -1
# That is ALWAYS truthy (or -1 doesn't do what they hoped — Python reads it
# as (player_hp == 0) or (-1), and -1 counts as True).
# Write it correctly: fatal is True when hp is 0 OR hp is -1... but the
# robust version every game uses is simply "hp is 0 or below".
# TODO:
fatal = ...

File: exercises/cpp/floor02/ex01_booleans.cpp
// FLOOR C2 · EXERCISE c2-1 — Booleans & comparisons
// Lesson: c2.1
//
// Same state checks as the Python wing — with C++'s own trap: = vs ==.
// `if (hp = 0)` ASSIGNS zero and evaluates false. It compiles. It ruins
// evenings. (Compile with -Wall and the compiler warns you — dojo does.)
//
// Run:  python3 dojo.py check c2-1
// Save: python3 dojo.py done c2-1

int player_hp = 7;
int player_max_hp = 30;
bool has_key = true;
bool door_locked = true;
int potions = 0;
int depth = 3;
bool torch_lit = false;

// hurt when below max
bool is_hurt()   { if (player_hp < player_max_hp) { return true; } return false; }

// at or below 25% of max. CAREFUL: player_max_hp / 4 in ints is 7 (30/4
// truncates). Here that happens to work — but write it robustly with a
// double: player_hp <= player_max_hp * 0.25
bool is_critical()  { if (double(player_hp) <= player_max_hp * 0.25) { return true; } return false; }

// has the key AND the door is locked
bool can_open_door() { if (has_key && door_locked) {return true; } return false; }

// critical AND out of potions
bool should_retreat() { if(potions == 0 && is_critical()){ return true; } return false; }

// torch NOT lit and deeper than floor 1
bool in_darkness()  { if (!torch_lit && depth > 1) { return true; } return false; }

// the robust death check: 0 or below (never ==)
bool fatal()        { if (player_hp <= 0) return true; else return false; }
